server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -2466,17 +2481,22 @@ protected void handleListeners(List<ActionListener<Void>> deleteListeners) {
      */
     private void failAllListenersOnMasterFailOver(Exception e) {
         logger.debug("Failing all snapshot operation listeners because this node is not master any longer", e);
+        final List<Runnable> readyToResolveListeners = new ArrayList<>();
         synchronized (currentlyFinalizing) {
             if (ExceptionsHelper.unwrap(e, NotMasterException.class, FailedToCommitClusterStateException.class) != null) {
                 repositoryOperations.clear();
-                for (Snapshot snapshot : Set.copyOf(snapshotCompletionListeners.keySet())) {
-                    failSnapshotCompletionListeners(snapshot, new SnapshotException(snapshot, "no longer master"));
+                for (final Snapshot snapshot : snapshotCompletionListeners.keySet()) {
+                    failSnapshotCompletionListeners(
+                        snapshot,
+                        new SnapshotException(snapshot, "no longer master"),
+                        readyToResolveListeners::add
+                    );
                 }
                 final Exception wrapped = new RepositoryException("_all", "Failed to update cluster state during repository operation", e);
```

Comment:
We should probably guard this one against unnecessary creation too - unlike the one in my previous comment this one is actually expensive (it computes the stack trace). That can be a follow-up tho, keeping this PR focussed just on the one change.

Reply:
I'll do this as a follow-up